temporal-sdk/src/main/java/io/temporal/internal/testservice/StateMachines.java

```diff
@@ -830,6 +833,9 @@ private static void startWorkflow(
     if (data.lastCompletionResult != null) {
       a.setLastCompletionResult(data.lastCompletionResult);
     }
+    if (data.lastFailure != null && data.lastFailure.isPresent()) {
```

Comment:
We shouldn't allow null values for Optional fields. They should be either Optiona.empty or populated.

Reply:
Maybe I'm still stuck in Rust land but, I worry about the idea that we can only rely on our own consistency to ensure Optionals are non-null. The null check is cheap and not too noisy, and allows us to consider the method by itself without ensuring `data` is always well constructed.

So, I agree with the theory, but in practice I like to check because it decreases mental load. Thoughts?

Reply:
Java has nice mechanism called NullPointerException to report such bugs. So adding repetitive checks all other the code for null is not serving any purpose besides adding clutter. In this particular case if  you want to add the check do in the constructor of Data.

 ```this.lastFailure = Objects.requireNotNull(lastFailure);```

I would also change all Data fields that are not expected to be changed to `final`.

Reply:
Ah, that's a nice way to handle it. Thanks.